blockstore/splitstore/splitstore.go

```diff
@@ -0,0 +1,831 @@
+		go func() {
+			defer atomic.StoreInt32(&s.compacting, 0)
+
+			log.Info("compacting splitstore")
+			start := time.Now()
+
+			s.compact()
+
+			log.Infow("compaction done", "took", time.Since(start))
+		}()
+	} else {
+		// no compaction necessary
+		atomic.StoreInt32(&s.compacting, 0)
+	}
+
+	return nil
+}
+
+// Compaction/GC Algorithm
+func (s *SplitStore) compact() {
+	if s.liveSetSize == 0 {
+		start := time.Now()
+		log.Info("estimating live set size")
+		s.estimateLiveSetSize()
+		log.Infow("estimating live set size done", "took", time.Since(start), "size", s.liveSetSize)
+	} else {
+		log.Infow("current live set size estimate", "size", s.liveSetSize)
+	}
+
+	if s.fullCompaction {
+		s.compactFull()
+	} else {
+		s.compactSimple()
+	}
+}
+
+func (s *SplitStore) estimateLiveSetSize() {
+	s.mx.Lock()
+	curTs := s.curTs
+	s.mx.Unlock()
+
+	s.liveSetSize = 0
+	err := s.cs.WalkSnapshot(context.Background(), curTs, 1, s.skipOldMsgs, s.skipMsgReceipts,
+		func(cid cid.Cid) error {
+			s.liveSetSize++
+			return nil
+		})
+
+	if err != nil {
+		// TODO do something better here
+		panic(err)
+	}
+}
+
+func (s *SplitStore) compactSimple() {
+	s.mx.Lock()
+	curTs := s.curTs
+	s.mx.Unlock()
+
+	coldEpoch := s.baseEpoch + CompactionCold
+
+	log.Infow("running simple compaction", "currentEpoch", curTs.Height(), "baseEpoch", s.baseEpoch, "coldEpoch", coldEpoch)
+
+	coldSet, err := s.env.NewLiveSet("cold", s.liveSetSize)
+	if err != nil {
+		// TODO do something better here
+		panic(err)
+	}
+	defer coldSet.Close() //nolint:errcheck
+
+	// 1. mark reachable cold objects by looking at the objects reachable only from the cold epoch
+	log.Info("marking reachable cold objects")
+	startMark := time.Now()
+
+	coldTs, err := s.cs.GetTipsetByHeight(context.Background(), coldEpoch, curTs, true)
+	if err != nil {
+		// TODO do something better here
+		panic(err)
+	}
+
+	count := int64(0)
+	err = s.cs.WalkSnapshot(context.Background(), coldTs, 1, s.skipOldMsgs, s.skipMsgReceipts,
+		func(cid cid.Cid) error {
+			count++
+			return coldSet.Mark(cid)
+		})
+
+	if err != nil {
+		// TODO do something better here
+		panic(err)
+	}
+
+	if count > s.liveSetSize {
+		s.liveSetSize = count
+	}
+
+	log.Infow("marking done", "took", time.Since(startMark))
+
+	// 2. move cold unreachable objects to the coldstore
+	log.Info("collecting cold objects")
+	startCollect := time.Now()
+
+	cold := make(map[cid.Cid]struct{})
+
+	// some stats for logging
+	var stHot, stCold int
+
+	// 2.1 iterate through the snoop and collect unreachable cold objects
+	err = s.snoop.ForEach(func(cid cid.Cid, wrEpoch abi.ChainEpoch) error {
+		// is the object stil hot?
+		if wrEpoch > coldEpoch {
+			// yes, stay in the hotstore
+			stHot++
+			return nil
+		}
+
+		// check whether it is reachable in the cold boundary
+		mark, err := coldSet.Has(cid)
+		if err != nil {
+			return xerrors.Errorf("error checkiing cold set for %s: %w", cid, err)
+		}
+
+		if mark {
```

Comment:
that's totally fine, we are salting the bloom filter with a random salt on every iteration so that we don't get the same false positives in every compaction.
Given that the false positives change, the garbage will get purged over time.